subsys/net/lib/nrf_cloud/coap/src/dtls.c

```diff
@@ -0,0 +1,325 @@
+			mfw_has_cid = false;
+		}
+	}
+
+	int verify = TLS_PEER_VERIFY_REQUIRED;
+
+	LOG_INF("  Peer verify: %d", verify);
+	err = setsockopt(sock, SOL_TLS, TLS_PEER_VERIFY, &verify, sizeof(verify));
+	if (err) {
+		LOG_ERR("Failed to setup peer verification, errno %d", -errno);
+		return -errno;
+	}
+
+	int session_cache = TLS_SESSION_CACHE_ENABLED;
+
+	err = setsockopt(sock, SOL_TLS, TLS_SESSION_CACHE, &session_cache, sizeof(session_cache));
+	if (err) {
+		LOG_ERR("Failed to enable session cache, errno: %d", -errno);
+		err = -errno;
+	}
+	return err;
+}
+
+bool dtls_cid_is_available(void)
+{
+	return mfw_has_cid;
+}
+
+int dtls_session_save(int sock)
+{
+	int dummy = 0;
+	int err;
+
+	LOG_DBG("Save DTLS CID session");
+	err = setsockopt(sock, SOL_TLS, TLS_DTLS_CONN_SAVE, &dummy, sizeof(dummy));
+	if (err) {
+		LOG_DBG("Failed to save DTLS CID session, errno %d", -errno);
+		err = -errno;
+	}
+	return err;
+}
+
+int dtls_session_load(int sock)
+{
+	int dummy = 1;
+	int err;
+
+	LOG_DBG("Load DTLS CID session");
+	err = setsockopt(sock, SOL_TLS, TLS_DTLS_CONN_LOAD, &dummy, sizeof(dummy));
+	if (err) {
+		LOG_DBG("Failed to load DTLS CID session, errno %d", -errno);
+		err = -errno;
+	}
+	return err;
+}
+
+bool dtls_cid_is_active(int sock)
+{
+	int err = 0;
+
+	if (dtls_cid_active) {
+		return true;
+	}
+
+	if (!mfw_has_cid) {
+		return false;
+	}
+
+	int status = 0;
+	int len = sizeof(status);
+
+	err = getsockopt(sock, SOL_TLS, TLS_DTLS_HANDSHAKE_STATUS, &status, &len);
+	if (!err) {
+		if (len > 0) {
+			if (status == TLS_DTLS_HANDSHAKE_STATUS_FULL) {
+				LOG_INF("Full DTLS handshake performed");
+			} else if (status == TLS_DTLS_HANDSHAKE_STATUS_CACHED) {
+				LOG_INF("Cached DTLS handshake performed");
+			} else {
+				LOG_WRN("Unknown DTLS handshake status: %d", status);
+			}
+		} else {
+			LOG_WRN("No DTLS status provided");
+		}
+	} else if ((errno != EOPNOTSUPP) && (errno != EINVAL)) {
+		LOG_ERR("Error retrieving handshake status: %d", -errno);
+	} /* else the current modem firmware does not support this feature */
+
+	len = sizeof(status);
+	err = getsockopt(sock, SOL_TLS, TLS_DTLS_CID_STATUS, &status, &len);
+	if (!err) {
+		if (len > 0) {
+			switch (status) {
+			case TLS_DTLS_CID_STATUS_DISABLED:
+				dtls_cid_active = false;
+				LOG_INF("No DTLS CID used");
+				break;
+			case TLS_DTLS_CID_STATUS_DOWNLINK:
+				dtls_cid_active = false;
+				LOG_INF("DTLS CID downlink");
+				break;
+			case TLS_DTLS_CID_STATUS_UPLINK:
+				dtls_cid_active = true;
+				LOG_INF("DTLS CID uplink");
+				break;
+			case TLS_DTLS_CID_STATUS_BIDIRECTIONAL:
+				dtls_cid_active = true;
+				LOG_INF("DTLS CID bidirectional");
+				break;
+			default:
+				LOG_WRN("Unknown DTLS CID status: %d", status);
+				break;
+			}
+		} else {
+			LOG_WRN("No DTLS CID status provided");
+		}
+	} else {
+		LOG_ERR("Error retrieving DTLS CID status: %d", -errno);
+	}
+
+	len = sizeof(status);
+	err = getsockopt(sock, SOL_TLS, TLS_DTLS_CID, &status, &len);
```

Comment:
could you add a brief comment on `TLS_DTLS_CID_STATUS` vs `TLS_DTLS_CID`?
it seems that `TLS_DTLS_CID_DISABLED` can override what is set by `TLS_DTLS_CID_STATUS`?

Reply:
The documentation for these features was unclear when I wrote this. I initially tried both methods to query the status.

The former indicates whether CID was successfully negotiated between client and server, and which direction provided a connection id (it can be one-sided or bidirectional).

The latter indicates what was set previously in dtls_setup().

I can remove the getsockopt of TLS_DTLS_CID at this point.